src/sqlparser/src/parser.rs

```diff
@@ -1242,46 +1211,66 @@ impl Parser {
-    /// if `named` is `true`, came from an expression like  `ARRAY[ex1, ex2]`
-    pub fn parse_array_expr(&mut self, named: bool) -> Result<Expr, ParserError> {
-        self.increase_array_depth(1);
-        if self.array_named_stack.len() < self.peek_array_depth() {
-            self.array_named_stack.push(named);
-        } else if let Err(parse_err) = self.check_same_named_array(named) {
-            Err(parse_err)?
-        }
-
-        if self.peek_token() == Token::RBracket {
-            let _ = self.next_token(); // consume ]
-            self.decrease_array_depth(1);
-            Ok(Expr::Array(Array {
+    pub fn parse_array_expr(&mut self) -> Result<Expr, ParserError> {
+        self.expect_token(&Token::LBracket)?;
+
+        if self.consume_token(&Token::RBracket) {
+            return Ok(Expr::Array(Array {
                 elem: vec![],
-                named,
-            }))
-        } else {
-            let exprs = self.parse_comma_separated(Parser::parse_expr)?;
-            self.expect_token(&Token::RBracket)?;
-            if self.array_named_stack.len() > self.peek_array_depth() {
-                self.array_named_stack.pop();
-            }
-            self.decrease_array_depth(1);
-            Ok(Expr::Array(Array { elem: exprs, named }))
+                named: true,
+            }));
         }
+
+        let mut expected_depth = None;
+        let exprs =
+            self.parse_comma_separated(|parser| parser.parse_array_inner(0, &mut expected_depth))?;
+        self.expect_token(&Token::RBracket)?;
+        Ok(Expr::Array(Array {
+            elem: exprs,
+            named: true,
+        }))
     }
 
-    fn check_same_named_array(&mut self, current_named: bool) -> Result<(), ParserError> {
-        let previous_named = self.array_named_stack.last().unwrap();
-        if current_named != *previous_named {
-            // for '['
-            self.prev_token();
-            if current_named {
-                // for keyword 'array'
-                self.prev_token();
-            }
-            parser_err!(format!("syntax error at or near {}", self.peek_token()))?
+    fn parse_array_inner(
+        &mut self,
+        depth: usize,
+        // All elements in the same array should have the same depth. When meet the first leaf element, set the value.
+        expected_depth: &mut Option<usize>,
+    ) -> Result<Expr, ParserError> {
+        Ok(if self.consume_token(&Token::LBracket) {
+            let exprs = self.parse_comma_separated(|parser| {
+                parser.parse_array_inner(depth + 1, expected_depth)
+            })?;
+            self.expect_token(&Token::RBracket)?;
+            Expr::Array(Array {
+                elem: exprs,
+                // Only top-level array is named.
+                named: false,
+            })
+        } else if self.peek_token() == Token::LBracket {
+            Expr::Array(Array {
+                elem: vec![],
+                named: false,
+            })
```

Comment:
What is this branch used for? I doubt it will never be entered? 🤔

Reply:
It's removed, the empty subarray is handled by another branch.